### PR TITLE
feat(sync): introduce Semaphore and SpinLock primitives

### PR DESCRIPTION
**Pull request by amery**
## Summary by CodeRabbit

- **New Features**
  - Introduced a lightweight spinlock for efficient mutual exclusion in low-contention scenarios.
  - Added a robust semaphore offering both exclusive and shared locking, with context-aware options for cancellation and timeouts.

- **Documentation**
  - Enhanced documentation with detailed sections for the new synchronization primitives, including `Semaphore` and `SpinLock`.

- **Tests**
  - Implemented comprehensive test suites for both `Semaphore` and `SpinLock`, covering various functionalities and edge cases.

### CODE REVIEW

**Comment by amery** [comment]
improved benchmarks and documentation

File: sync/spinlock/spinlock_test.go [hunk 1]
```diff
@@ -0,0 +1,311 @@
...
+
+func BenchmarkContention_StdMutex(b *testing.B) {
+	var mu sync.Mutex
+	n := 1
+
+	b.RunParallel(func(pb *testing.PB) {
+		for pb.Next() {
+			mu.Lock()
+			// Simulate work with some CPU-bound operations
+			for range 50 {
```
[the remark this answers is not in the document]
Comment (amery) [reply sync/spinlock/spinlock_test.go:247]: `for range N` was introduced in Go 1.22. The minimum version we support

File: sync/spinlock/spinlock_test.go [hunk 2]
```diff
@@ -0,0 +1,311 @@
...
+// Contention benchmarks
+func BenchmarkContention_SpinLock(b *testing.B) {
+	var sl SpinLock
+	n := 1
+
+	b.RunParallel(func(pb *testing.PB) {
+		for pb.Next() {
+			sl.Lock()
+			// Simulate work with some CPU-bound operations
+			for range 50 {
```
[the remark this answers is not in the document]
Comment (amery) [reply sync/spinlock/spinlock_test.go:231]: @coderabbitai `for range N` loops were introduced in Go 1.22. The minimum version we support

File: sync/semaphore/semaphore_test.go [hunk 3]
```diff
@@ -0,0 +1,1062 @@
...
+	})
+
+	t.Run("sequential locks", func(_ *testing.T) {
+		s := &Semaphore{}
+		for range 5 {
+			s.Lock()
+			doSomething()
+			s.Unlock()
+		}
+	})
```
[the remark this answers is not in the document]
Comment (amery) [reply sync/semaphore/semaphore_test.go:66]: we use go 1.22 as the minimum version @coderabbitai

**Comment by amery** [comment]
drop Semaphore's Acquire/Release methods in favour of LockContext/Unlock

**Comment by amery** [comment]
work to add writer starvation prevention to the Semaphore is underway, but not critical at this stage
